fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1529,58 +1560,45 @@ def _free_full_params(self, params: Optional[List[Parameter]] = None) -> None:
-            # on how to split the "merged" parameters, by extracting the meta-data
-            # used in the FlattenParamsWrapper
-            else:
-                param_names = []
-                for module_path, param_name in m.param_path_infos:
-                    full_param_path = module_path + "." + param_name if module_path else param_name
-                    param_names.append(_clean_path(full_param_path))
-                params_metadata.append(
-                    {
-                        "fsdp_path": _clean_path(path),
-                        "is_flat": True,
-                        "num_padded": m.numel_padded_per_param,
-                        "param_names": param_names,
-                        # TODO (Min): we don't want to access the private _param_shapes and
-                        # _param_numels here. We want to dump metadata from FPW when there are
-                        # multiple groups of params.
-                        "param_shapes": m._fsdp_wrapped_module.flat_param._param_shapes,  # type: ignore
-                        "param_numels": m._fsdp_wrapped_module.flat_param._param_numels,  # type: ignore
-                        "no_broadcast_optim_state": m.no_broadcast_optim_state,
+            if isinstance(m, FullyShardedDataParallel):
+                metadata: Dict[str, Any] = {}
+                metadata["fsdp_path"] = _clean_path(path)
+                metadata["params"] = {}
+
+                metadata["no_broadcast_optim_state"] = m.no_broadcast_optim_state
+                shared_param_info = []
+                for (mpath_dst, mpath_src, _, src_name, _, dst_name) in m._shared_param_infos:
+                    src_param_path = _clean_path(mpath_src + "." + src_name if mpath_src else src_name)
+                    dst_param_path = _clean_path(mpath_dst + "." + dst_name if mpath_dst else dst_name)
+                    shared_param_info.append((src_param_path, dst_param_path))
+                metadata["shared_param_info"] = shared_param_info
+
+                for i, p in enumerate(m.params):
+                    if i < m._num_flatten_params:
+                        backing_param_name = m.module.flat_param_names[i]
+                        names, shapes, numels = m.module.metadata(i)
+                    else:
+                        assert len(m._param_name_groups[i]) == 1
+                        backing_param_name = m._param_name_groups[i][0]
+                        names = [backing_param_name]
+                        shapes = [p._orig_size]
+                        numels = [p._orig_size.numel()]
+                    backing_param_name = _clean_path(backing_param_name)
+                    metadata["params"][backing_param_name] = {
+                        "names": [_clean_path(n) for n in names],
+                        "shapes": shapes,
+                        "numels": numels,
+                        "padding_of_backing": m.numel_padded_per_param[i],
```

Comment:
I would just call this padding

the of_backing names make sense internally but not to a random caller reading a returned dictionary.